src/legacy/reduce.rs

```diff
@@ -0,0 +1,488 @@
+        0x1a, 0x5e, 0x48, 0x0a, 0x85, 0x4c, 0xc0, 0x0d, 0x39, 0xa1, 0x1b, 0x52, 0x28, 0xec, 0xac,
+        0xf0, 0x13, 0x52, 0x06, 0xa4, 0x42, 0x0a, 0xc1, 0x14, 0x24, 0x17, 0x7c, 0x04, 0x81, 0x44,
+        0x23, 0x9b, 0x29, 0x07, 0x20, 0x2c, 0x0f, 0x42, 0x90, 0xd0, 0xee, 0x06, 0x87, 0x96, 0x42,
+        0x8a, 0x42, 0x4a, 0x2b, 0x64, 0x63, 0x12, 0x52, 0x14, 0x84, 0x9c, 0x71, 0x0a, 0x29, 0x11,
+        0x27, 0x94, 0x68, 0x84, 0x43, 0xd3, 0x00, 0xa3, 0xd4, 0x88, 0x96, 0x71, 0x9b, 0x20, 0x82,
+        0x43, 0xb6, 0x58, 0x85, 0xec, 0x02, 0x33, 0xc1, 0x8a, 0x15, 0x42, 0x71, 0x69, 0x85, 0x3c,
+        0xfc, 0x42, 0x1e, 0xa9, 0x86, 0xbc, 0xf1, 0x30, 0xe6, 0x75, 0xe5, 0x8e, 0x79, 0xde, 0x30,
+        0x24, 0x13, 0x4b, 0x6c, 0x42, 0x0e, 0x3b, 0x96, 0xa8, 0xdc, 0xb0, 0x6d, 0x6a, 0x1a, 0x81,
+        0x65, 0x3a, 0xf7, 0x4d, 0x87, 0x4d, 0x21, 0x87, 0xc5, 0x83, 0x6c, 0x13, 0x28, 0x67, 0x20,
+        0x8a, 0x6d, 0xe3, 0xc1, 0xfb, 0x50, 0x26, 0xab, 0x9c, 0x54, 0x75, 0x8a, 0x85, 0x4b, 0x0c,
+        0x62, 0x87, 0x7c, 0xb0, 0xc1, 0x62, 0xb2, 0xd1, 0x90, 0x45, 0xc4, 0x15, 0xa2, 0xcc, 0x0f,
+        0xa4, 0x62, 0x1f, 0x21, 0x31, 0x45, 0x15, 0x72, 0x59, 0xba, 0x6c, 0xc4, 0x98, 0xb5, 0x34,
+        0x10, 0x15, 0xba, 0x34, 0x1b, 0x16, 0x72, 0x58, 0x4f, 0x17, 0x79, 0x54, 0x04, 0x5c, 0xa5,
+        0x59, 0x2c, 0x66, 0x54, 0xdd, 0xb2, 0x65, 0x84, 0x0a, 0xaf, 0xda, 0x28, 0xf6, 0x98, 0x85,
+        0x6e, 0xf2, 0x2e, 0x08, 0xa8, 0x59, 0xc8, 0x72, 0x13, 0x86, 0xb2, 0x69, 0x9d, 0x69, 0x74,
+        0x11, 0x9f, 0x98, 0x3e, 0x39, 0x85, 0x74, 0x4e, 0xa6, 0x6f, 0x48, 0x86, 0x43, 0x10, 0x72,
+        0xd4, 0x0d, 0xa4, 0xd1, 0xba, 0x48, 0x26, 0x8b, 0x60, 0xd1, 0x29, 0x16, 0xe8, 0x4d, 0x30,
+        0x2a, 0x1d, 0x72, 0xcd, 0xa4, 0x8b, 0x7c, 0x82, 0x42, 0x32, 0xd3, 0xa4, 0x20, 0x16, 0x12,
+        0xb1, 0xee, 0x59, 0xb4, 0x90, 0xa3, 0x26, 0x20, 0x2f, 0x7c, 0x20, 0x21, 0x25, 0x95, 0x9f,
+        0x58, 0x68, 0x24, 0xe7, 0x65, 0x34, 0x0d, 0x7b, 0xc2, 0xb9, 0xbe, 0x2e, 0xd2, 0xe8, 0x49,
+        0x0a, 0x3b, 0x29, 0xe5, 0x14, 0xe4, 0x0c, 0x18, 0x27, 0x00,
+    ];
+
+    #[test]
+    fn test_expand_hamlet2048() {
+        let mut dst = VecDeque::new();
+        let mut src_used = 0;
+        hwexpand(
+            &HAMLET_2048,
+            HAMLET_2048.len(),
+            2048,
+            4,
+            &mut src_used,
+            &mut dst,
+        )
+        .unwrap();
+        assert_eq!(dst.len(), 2048);
+    }
+
+    /*
+      Put some text first to make PKZIP actually use Reduce compression.
+      Target the code path which copies a zero when dist > current position.
+
+      $ curl -O http://cd.textfiles.com/originalsw/25/pkz092.exe
+      $ dosbox -c "mount c ." -c "c:" -c "pkz092" -c "exit"
+      $ dd if=hamlet.txt bs=1 count=2048 > a
+      $ dd if=/dev/zero  bs=1 count=1024 >> a
+      $ dosbox -c "mount c ." -c "c:" -c "pkzip -ea4 a.zip a" -c "exit"
+      $ xxd -i -s 31 -l $(expr $(find A.ZIP -printf %s) - 100) A.ZIP
+    */
+    const ZEROS_REDUCED: [u8; 1297] = [
```

Comment:
Move this to a file as well, or else start with [0u8; 1297].

Reply:
moved.